### PR TITLE
feat(prob): support protocol probing

### PR DESCRIPTION
**Pull request by whalecold**
#### What type of PR is this?

feat
#### Check the PR title.

- [x] This PR title match the format: \<type\>(optional scope): \<description\>
- [x] The description of this PR title is user-oriented and clear enough for others to understand.
- [x] Attach the PR updating the user documentation if the current PR requires user awareness at the usage level. [User docs repo](https://github.com/cloudwego/cloudwego.github.io)

#### (Optional) Translate the PR title into Chinese.
支持协议嗅探，同时支持 dubbo 和原生的 thrift

#### (Optional) More detailed description for this PR(en: English/zh: Chinese).

en:
zh(optional): 

#### (Optional) Which issue(s) this PR fixes:

https://github.com/kitex-contrib/codec-dubbo/issues/91

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -323,6 +323,39 @@ service EchoService {
...
+	"net"
+)
+
+func main() {
+	// 指定服务端将要监听的地址
+	addr, _ := net.ResolveTCPAddr("tcp", ":21000")
+	svr := hello.NewServer(new(GreetServiceImpl),
+		server.WithServiceAddr(addr),
+		// 配置 DubboCodec
+		server.WithTransHandlerFactory(dubbo.NewSvrTransHandlerFactory(
```
Comment (DMwangnima) [inline README.md]: I think we should use
```
detection.NewSvrTransHandlerFactory(netpoll.NewSrvTransHandlerFactory, 
dubbo.NewSvrTransHandlerFactory(),
nphttp2.NewSvrTransHandlerFactory(),
)
```
to inject it.
Refer to https://github.com/cloudwego/kitex/blob/a74dfc289a3805d65f7345257d80743c1f2b6523/internal/server/remote_option.go#L35.

Comment (whalecold) [reply README.md]: Done

File: pkg/server_trans_handler_factory.go [hunk 2]
```diff
@@ -0,0 +1,98 @@
+/*
+ * Copyright 2023 CloudWeGo Authors
```
Comment (DMwangnima) [inline pkg/server_trans_handler_factory.go]: 2024

File: pkg/server_trans_handler_factory.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	npReader := conn.(interface{ Reader() cnetpoll.Reader }).Reader()
+	// read at most avoid block
+	header, err := npReader.Peek(dubbo_spec.HEADER_SIZE)
+	if err != nil {
+		return err
+	}
+	if header[0] == dubbo_spec.MAGIC_HIGH && header[1] == dubbo_spec.MAGIC_LOW {
+		return nil
+	}
+	return errors.New("error protocol not match hessian")
```
Comment (DMwangnima) [inline pkg/server_trans_handler_factory.go]: "error protocol not match dubbo"

File: README_ENG.md [hunk 4]
```diff
@@ -318,6 +318,43 @@ service EchoService {
...
+	hello "demo-server/kitex_gen/hello/greetservice"
+)
+
+func main() {
+	addr, _ := net.ResolveTCPAddr("tcp", ":21000")
+	svr := hello.NewServer(new(GreetServiceImpl),
+		server.WithServiceAddr(addr),
+		server.WithTransHandlerFactory(detection.NewSvrTransHandlerFactory(netpoll.NewSvrTransHandlerFactory(),
+			dubbo.NewSvrTransHandlerFactory(
+				// 配置 Kitex 服务所对应的 Java Interface. 其他 dubbo 客户端和 kitex 客户端可以通过这个名字进行调用。
```
Comment (DMwangnima) [inline README_ENG.md]: Pls change to english version.

**Review by DMwangnima — COMMENTED** [review]
BTW. Could you add crosstest to https://github.com/kitex-contrib/codec-dubbo-tests ?

File: pkg/server_trans_handler_factory.go (hunk 2 above)
Comment (whalecold) [reply pkg/server_trans_handler_factory.go]: image
it reports error when use 2024 .

Comment (DMwangnima) [reply pkg/server_trans_handler_factory.go]: The quickest way is to modify the pattern in ```.licenserc.yaml```.
Please refer to https://github.com/kitex-contrib/codec-dubbo-tests/blob/845960f06eef4e5ad0a0d26bf589e265ca65c325/.licenserc.yaml#L24

**Comment by whalecold** [comment]
> BTW. Could you add crosstest to https://github.com/kitex-contrib/codec-dubbo-tests ?

Ok, I will do that later.

**Review by DMwangnima — APPROVED** [review]

**Pull request merged**